### PR TITLE
Wasm server integration

### PR DESCRIPTION
**Pull request by patrick-east**
There are few things going on with these changes:

* Adding a new "Resolver" interface and API plumbed through for evaluation which allows arbitrary ref's to have values "resolved" by external things from the topdown evaluation or storage lookup.
* Adding an implementation of the Resolver interface for evaluating OPA Wasm modules using the golang-opa-wasm sdk.
* Adding support for bundles to include >1 Wasm module and provide a mapping of entrypoints to modules in the manifest.
* When configured as such a wasm resolver will be instantiated and used by the opa server.
* Moving the golang-opa-wasm SDK into the opa repo
* Updates to the SDK to support multiple entrypoints

### CODE REVIEW

File: cmd/build.go [hunk 1]
```diff
@@ -107,8 +107,8 @@ The 'build' command supports targets (specified by -t):
             files. If optimization is enabled at least one entrypoint (-e) must be supplied.
 
     wasm    The wasm target emits a bundle containing a WebAssembly module compiled from
-            the input files. The bundle may contain the original policy or data files.
-            The wasm target requires exactly one entrypoint (-e) be supplied.
+            the input files for each specified entrypoint. The bundle may contain the
+		    original policy or data files.
```
Comment (tsandall) [inline cmd/build.go:112]: Indentation on this looks odd.

File: bundle/store.go [hunk 2]
```diff
@@ -6,41 +6,52 @@ package bundle
...
+	"encoding/base64"
+	"encoding/json"
 	"fmt"
 	"strings"
 
-	"github.com/open-policy-agent/opa/metrics"
-
 	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/metrics"
+	"github.com/open-policy-agent/opa/resolver/wasm"
```
Comment (tsandall) [inline bundle/store.go:16]: The bundle package was originally meant to be a simple, reusable package for interacting with bundles--over time the package has taken on dependencies with the OPA in-memory store and now we're adding a dependency on the wasm runtime. I'm a bit worried about the scope creep here. It means that anyone who wants to simply parse bundles to inspect the files/manifests will need to pull in the wasm runtime. If there's some way we can avoid these dependencies, I think it would be better.

I'd also like to see the bundle activation logic moved into a separate package because that's really an implementation detail of how we do file loading in OPA but we can solve that another time.

File: test/authz/testing.go [hunk 3]
```diff
@@ -15,7 +15,7 @@ import (
 )
 
 // Policy is a test rego policy for a token based authz system
-const Policy = `package restauthz
+const Policy = `package policy.restauthz
```
Comment (tsandall) [inline test/authz/testing.go:18]: Just curious why this change was needed.

File: test/e2e/wasm/authz/authz_bench_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	runAuthzBenchmark(b, testAuthz.Allow, 10)
+}
+
+func BenchmarkRESTAuthzAllow100Paths(b *testing.B) {
+	runAuthzBenchmark(b, testAuthz.Allow, 100)
+}
+
+func BenchmarkRESTAuthzAllow1000Paths(b *testing.B) {
+	runAuthzBenchmark(b, testAuthz.Allow, 1000)
+}
```
Comment (tsandall) [inline test/e2e/wasm/authz/authz_bench_integration_test.go:118]: It would be nice to include the output from go test -bench in the commit message for these so we have a rough sense of what we're looking at.

File: bundle/bundle.go [hunk 5]
```diff
@@ -44,11 +44,12 @@ const (
...
-	Manifest   Manifest
-	Data       map[string]interface{}
-	Modules    []ModuleFile
-	Wasm       []byte
+	Signatures  SignaturesConfig
+	Manifest    Manifest
+	Data        map[string]interface{}
+	Modules     []ModuleFile
+	Wasm        []byte // Deprecated. Use WasmModules instead
+	WasmModules []WasmModuleFile
```
Comment (tsandall) [inline bundle/bundle.go:52]: Since we've added support for multiple entrypoints inside of a single .wasm file I do wonder whether it's worth the complexity of supporting multiple modules inside the bundle. The cost of instantiating multiple VMs is high enough that we won't use this for the foreseeable future. Of course, moving from 1->N in the future would be harder so maybe we should just bite the bullet now. Just something to consider.

**Review by tsandall — COMMENTED** [review]
Left a few comments. Overall this looks great. Nice job dealing w/ all the plumbing. I'm going to pull down the changes and try them out now.

File: bundle/bundle.go (hunk 5 above)
Comment (patrick-east) [reply bundle/bundle.go:52]: Yea I have been eyeing that issue as I got this rebased on the changes to support >1 entrypoint per wasm module. Initially I was thinking to remove it... but I tend to agree that the cost of adding it later on would likely be higher than supporting it out of the gate. I still need to make the changes to have each wasm resolver support multiple entrypoints, will see if anything in that process changes my mind on it 😄

File: test/e2e/wasm/authz/authz_bench_integration_test.go (hunk 4 above)
Comment (patrick-east) [reply test/e2e/wasm/authz/authz_bench_integration_test.go:118]: Will do, ill add the output and a comparison against the similar test using the normal rego evaluations.

File: test/authz/testing.go (hunk 3 above)
Comment (patrick-east) [reply test/authz/testing.go:18]: If I remember correctly it was because we needed to push data into the store as part of the tests, like each test has a set of "external data" plus the common policy. To allow the tests to do a PUT to /v1/data/whatever and use bundles I moved the policies in the bundle into their own sub-path and defined roots for the bundle specific to them.

File: bundle/store.go (hunk 2 above)
Comment (patrick-east) [reply bundle/store.go:16]: I'll look into refactoring some of that, IIRC some of this was just out of convenience.

File: cmd/build.go (hunk 1 above)
Comment (patrick-east) [reply cmd/build.go:112]: 🤔  I think my file watcher thingy got disabled at some point, will fix (and some of the other spots that are failing the lint check)

File: bundle/bundle.go [hunk 6]
```diff
@@ -89,8 +90,15 @@ func NewFile(name, hash, alg string) FileInfo {
...
-	Revision string    `json:"revision"`
-	Roots    *[]string `json:"roots,omitempty"`
+	Revision      string         `json:"revision"`
+	Roots         *[]string      `json:"roots,omitempty"`
+	WasmResolvers []WasmResolver `json:"wasm,omitempty"`
+}
+
+// WasmResolver maps a wasm module to an entrypoint ref.
+type WasmResolver struct {
+	Entrypoint string `json:"entrypoint,omitempty"`
```
[the remark this answers is not in the document]
Comment (patrick-east) [reply bundle/bundle.go:100]: Yep, it is the same "entrypoint" as like the compile package (and `opa build`) use. I had figured to keep the naming similar across them and standardize on an "entrypoint" being a non-data prefixed ref path (eg `authz/allow` or whatever).

I've been thinking about that, whether or not we should have separate "entrypoint" and like "query name" fields on these in the manifest (or in the stored stuff). When using `opa build` or the compile package they will always be the same, presumably anyone hand-crafting bundles or compiling wasm could do whatever they wanted... but I'm not certain if we need/want to support that initially.

File: cmd/build.go (hunk 1 above)
Comment (patrick-east) [reply cmd/build.go]: Fixed

**Comment by patrick-east** [comment]
^ the latest changes are still much of a WIP, they are not ready to test yet, only pushed for review purposes. Will update when ready to take for a spin

**Comment by patrick-east** [comment]
Ok, should be good to go for testing and review

File: topdown/resolver.go [hunk 7]
```diff
@@ -66,6 +66,7 @@ func (t *resolverTrie) mktree(ctx context.Context, in resolver.Input) (ast.Value
 	}
 	obj := ast.NewObject()
 	for k, child := range t.children {
+		in = resolver.Input{Ref: append(in.Ref.Copy(), ast.NewTerm(k)), Input: in.Input}
```
Comment (patrick-east) [inline topdown/resolver.go:69]: @tsandall curious if you have any suggestions on this, I'm not 100% sure the original intent of `Input.Ref`, if it was supposed to be the original one the resolver trie is trying to resolve or the ref that the individual resolver was being asked to resolve. Doing this copy/append business doesn't seem ideal, we could front load it when building the trie maybe?

File: bundle/bundle.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (patrick-east) [reply bundle/bundle.go:100]: Discussed offline, we will keep it as a single field named "entrypoint" for now.

File: bundle/store.go (hunk 2 above)
Comment (patrick-east) [reply bundle/store.go]: Latest changes move the helper to instantiates the wasm resolvers into an internal bundle utils package.

File: compile/compile.go [hunk 8]
```diff
@@ -430,6 +427,45 @@ func (c *Compiler) compileWasm(ctx context.Context) error {
...
+			}
+
+			// If any rules were dropped update the module accordingly
+			if len(rules) != len(mf.Parsed.Rules) {
+				mf.Parsed.Rules = rules
+				// Remove the original raw source, we're editing the AST
+				// directly so it wont be in sync anymore.
+				mf.Raw = nil
+			}
+		}
```
Comment (tsandall) [inline compile/compile.go:466]: These changes will leave the bundle in an invalid state if there are unresolved references to the entrypoint. For example:

example.rego:

```rego
package foo

p { q > 7 }
q = 6
```

If the entrypoint was `foo/q` then the resulting example.rego file in the output bundle would be:

```rego
package foo
p { q > 7}
```

This version of `p` is unsafe because `q` is not defined anywhere. We could either resolve the refs when we remove the rules or we could modify the ast.Compiler to resolve refs based on externs. I prefer the former because otherwise we're allowing for .rego files emitted by the compiler to be invalid on their own.

Comment (patrick-east) [reply compile/compile.go]: Updated in latest changes to inject the import statements

File: internal/wasm/sdk/opa/pool.go [hunk 9]
```diff
@@ -46,7 +48,10 @@ func newPool(poolSize, memoryMinPages, memoryMaxPages uint32) *pool {
 // Acquire obtains a VM from the pool, waiting if all VMms are in use
 // and building one as necessary. Returns either ErrNotReady or
 // ErrInternal if an error.
-func (p *pool) Acquire(ctx context.Context) (*vm, error) {
+func (p *pool) Acquire(ctx context.Context, metrics metrics.Metrics) (*vm, error) {
+	metrics.Timer("opa_wasm_pool_acquire_vm").Start()
+	defer metrics.Timer("opa_wasm_pool_acquire_vm").Stop()
```
Comment (tsandall) [inline internal/wasm/sdk/opa/pool.go:53]: nit; Let's drop the `opa_` prefix on the metric keys. We don't do that anywhere else.

File: internal/wasm/sdk/opa/vm.go [hunk 10]
```diff
@@ -170,8 +170,14 @@ func newVM(policy []byte, data []byte, memoryMin, memoryMax uint32) (*vm, error)
...
 
-func (i *vm) Eval(ctx context.Context, entrypoint EntrypointID, input *interface{}) (interface{}, error) {
-	if err := i.setHeapState(i.evalHeapPtr); err != nil {
+func (i *vm) Eval(ctx context.Context, entrypoint EntrypointID, input *interface{}, metrics metrics.Metrics) (interface{}, error) {
+	metrics.Timer("opa_wasm_vm_eval_total").Start()
+	defer metrics.Timer("opa_wasm_vm_eval_total").Stop()
+
+	metrics.Timer("opa_wasm_vm_eval_set_heap_state").Start()
+	err := i.setHeapState(i.evalHeapPtr)
+	metrics.Timer("opa_wasm_vm_eval_set_heap_state").Stop()
```
Comment (tsandall) [inline internal/wasm/sdk/opa/vm.go:179]: nit: I'd like to just see this bundled into the "input preparation" metric since we already have several calls into wasm land and this is just one more.

File: internal/wasm/sdk/opa/vm.go [hunk 11]
```diff
@@ -170,8 +170,14 @@ func newVM(policy []byte, data []byte, memoryMin, memoryMax uint32) (*vm, error)
 	return v, nil
 }
 
-func (i *vm) Eval(ctx context.Context, entrypoint EntrypointID, input *interface{}) (interface{}, error) {
-	if err := i.setHeapState(i.evalHeapPtr); err != nil {
+func (i *vm) Eval(ctx context.Context, entrypoint EntrypointID, input *interface{}, metrics metrics.Metrics) (interface{}, error) {
+	metrics.Timer("opa_wasm_vm_eval_total").Start()
```
Comment (tsandall) [inline internal/wasm/sdk/opa/vm.go:174]: Looking at the metric keys, here's what I'd like to see:

```
wasm_pool_acquire
wasm_pool_release # not sure how easy this one would be to add given it's used in SetData...
wasm_vm_eval
wasm_vm_eval_prepare_input
wasm_vm_eval_execute
wasm_vm_eval_prepare_result
```

File: resolver/wasm/wasm.go [hunk 12]
```diff
@@ -80,6 +80,8 @@ func (r *Resolver) Close() {
 // Eval performs an evaluation using the provided input and the Wasm module
 // associated with this Resolver instance.
 func (r *Resolver) Eval(ctx context.Context, input resolver.Input) (resolver.Result, error) {
+	input.Metrics.Timer("wasm_resolver_eval").Start()
+	defer input.Metrics.Timer("wasm_resolver_eval").Stop()
```
Comment (tsandall) [inline resolver/wasm/wasm.go:84]: This metric seems a bit redundant. I think we're essentially measuring the time to convert the AST input value into a native go value. Along those lines, I'm tempted to just pass AST input value into the SDK for two reasons:

- We'll need to do this anyway to support sets in the input
- We'll avoid the need to roundtrip the input value (in theory we could serialize from AST input value directly into the Wasm address space--in the short term we can just call `String()` and copy the bytes.)

**Review by tsandall — COMMENTED** [review]
Left a couple comments on metric naming.

File: compile/compile.go [hunk 13]
```diff
@@ -437,6 +437,10 @@ func (c *Compiler) compileWasm(ctx context.Context) error {
 
 	// Each entrypoint needs an entry in the manifest along with the
 	// original rule(s) removed from the remaining rego modules.
+
+	// For each package path keep a list of new imports. They are stored as a
```
Comment (tsandall) [inline compile/compile.go:441]: I think this logic is complex enough that we should factor it into a separate function.

Comment (tsandall) [reply compile/compile.go:441]: If there are multiple entrypoints in the same package the output will be non-deterministic since the imports are stored in a map. I could imagine just refactoring this to be map[string][]*ast.Import where the key is the package. The code below that adds the required imports can just search the slice to see if the import already exists (the slice should be small in practice).

Comment (tsandall) [reply compile/compile.go:441]: We should have a test case covering this...

Something like:

module1.rego

```
package test

p { q }
```

module2.rego:

```
package test

q { r }
r { s }
s = 7
```

Compile out entrypoints `q` and `s`. The result should be:

module1.rego:

```
package test

import data.test.q
import data.test.s

p { q }
```

module2.rego

```
package test

import data.test.q
import data.test.s

r { s }
```

File: topdown/resolver.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	node.r = r
+}
+
+func (t *resolverTrie) Resolve(e *eval, ref ast.Ref) (ast.Value, error) {
+	e.metrics.Timer(metrics.RegoExternalResolve).Start()
+	defer e.metrics.Timer(metrics.RegoExternalResolve).Stop()
+
+	in := resolver.Input{
+		Ref:     ref,
```
Comment (tsandall) [inline topdown/resolver.go:40]: This needs to change so that we only pass the entrypoint ref--not the entire ref. We can set the ref below:

```
in.Ref = ref[:i+1]
result, err := node.r.Eval(e.ctx, in)
```

Comment (tsandall) [reply topdown/resolver.go:40]: Sent a PR against the fork to fix this: https://github.com/patrick-east/opa/pull/21

File: internal/wasm/sdk/opa/pool.go (hunk 9 above)
Comment (patrick-east) [reply internal/wasm/sdk/opa/pool.go:53]: Will do. The prefix was just from the package name 😄  but I do like a "less is more" approach to these.

File: resolver/wasm/wasm.go (hunk 12 above)
Comment (patrick-east) [reply resolver/wasm/wasm.go:84]: > This metric seems a bit redundant

+1, It was a remnant of the first pass at plumbing these through to only the resolver, but after pushing them all the way down into the wasm sdk stuff it doesn't make sense anymore.

> I'm tempted to just pass AST input value into the SDK for two reasons

Sounds good to me, making the changes now.

Comment (patrick-east) [reply resolver/wasm/wasm.go]: Latest commit includes changes to support rego types in and out of the wasm evaluation. We're using the input stringer and copying bytes on the way in, then `opa_value_dump` and `ast.ParseTerm` so we can get set values on the way out.

File: internal/wasm/sdk/opa/config.go [hunk 15]
```diff
@@ -30,7 +30,7 @@ func (o *OPA) WithPolicyBytes(policy []byte) *OPA {
 	return o
 }
 
-// WithDataFile configures the JSON data file to load.
+// WithDataFile configures the JSON parsedData file to load.
```
Comment (tsandall) [inline internal/wasm/sdk/opa/config.go:33]: nit: looks like `s/data/parsedData/g` got run a bit too eagerly. I know it's super minor but it would be good to fix the comments and the error messages in internal/wasm/sdk/opa/errors.go

File: internal/wasm/sdk/opa/opa.go [hunk 16]
```diff
@@ -44,7 +44,7 @@ type EntrypointID int32
 // initialized before invoking the Eval.
 func New() *OPA {
 	opa := &OPA{
-		memoryMinPages: 2,
+		memoryMinPages: 16,
```
Comment (tsandall) [inline internal/wasm/sdk/opa/opa.go:47]: Should we just take `len(policyBytes)` and round to the nearest page size multiple?

File: internal/wasm/sdk/opa/pool.go [hunk 17]
```diff
@@ -129,14 +129,22 @@ func (p *pool) SetPolicyData(policy []byte, data []byte) error {
...
+			var vm *vm
+			vm, err = newVM(policy, data, p.memoryMinPages, p.memoryMaxPages)
+			if err == nil {
+				p.vms = append(p.vms, vm)
+				p.acquired = append(p.acquired, false)
+			} else {
+				err = fmt.Errorf("%v: %w", err, ErrInvalidPolicyOrData)
+				break
+			}
+		}
```
Comment (tsandall) [inline internal/wasm/sdk/opa/pool.go]: On my desktop, startup for a trivial bundle is several seconds (because this has to initialize one VM per CPU thread...) I'm tempted to just throw goroutines at this to make startup more response. WDYT?

File: internal/wasm/sdk/opa/opa.go [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+func (o *OPA) Entrypoints(ctx context.Context) (map[string]EntrypointID, error) {
+	instance, err := o.pool.Acquire(ctx, metrics.New())
+	if err != nil {
+		return nil, err
+	}
+
+	defer o.pool.Release(instance, metrics.New())
+
+	return instance.Entrypoints(), nil
+}
```
Comment (tsandall) [inline internal/wasm/sdk/opa/opa.go:215]: Do we need to expose this outside of the SDK? I was thinking the caller would just supply the entrypoint name and the SDK would be responsible for mapping to the internal ID (which is an implementation detail of the compiler).

File: internal/wasm/sdk/opa/vm.go [hunk 19]
```diff
@@ -113,8 +124,30 @@ func newVM(policy []byte, data []byte, memoryMin, memoryMax uint32) (*vm, error)
 		return nil, err
 	}
 
-	if data != nil {
-		if v.dataAddr, err = v.toRegoJSON(data, true); err != nil {
+	// Optimization for cloning a vm, if provided a parsed data memory buffer
```
Comment (tsandall) [inline internal/wasm/sdk/opa/vm.go:127]: Nice! With this optimization we only have to parse the entire data set once as opposed to once per core? LGTM--though we'll want to get test coverage ASAP.

File: internal/wasm/sdk/opa/vm.go [hunk 20]
```diff
@@ -548,8 +548,13 @@ func (i *vm) fromRegoJSON(addr int32, free bool) (interface{}, error) {
...
 // toRegoJSON converts go native JSON to Rego JSON.
 func (i *vm) toRegoJSON(v interface{}, free bool) (int32, error) {
-	raw, ok := v.([]byte)
-	if !ok {
+	var raw []byte
+	switch v := v.(type) {
+	case []byte:
+		raw = v
+	case ast.Value:
+		raw = []byte(v.String())
```
Comment (tsandall) [inline internal/wasm/sdk/opa/vm.go:556]: I'd add support for *ast.Term here as well.

File: server/server.go [hunk 21]
```diff
@@ -1388,6 +1400,12 @@ func (s *Server) v1DataPost(w http.ResponseWriter, r *http.Request) {
 			rego.StrictBuiltinErrors(strictBuiltinErrors),
 		}
 
+		for _, r := range s.manager.GetWasmResolvers() {
```
Comment (tsandall) [inline server/server.go:1403]: When a new bundle downloads and activates do these get invalidated by the server's reload function?

Comment (tsandall) [reply server/server.go:1403]: I assume we're doing this for performance? A short comment explaining why this is done would be useful.

File: plugins/plugins.go [hunk 22]
```diff
@@ -633,19 +633,22 @@ func requiresWasmResolverReload(event storage.TriggerEvent) bool {
...
-		for _, resolver := range m.wasmResolvers {
-			err := resolver.SetData(data)
+	for _, resolver := range m.wasmResolvers {
+		for _, dataEvent := range event.Data {
+			var err error
+			if dataEvent.Removed {
+				err = resolver.RemoveDataPath(dataEvent.Path)
+			} else {
+				err = resolver.SetDataPath(dataEvent.Path, dataEvent.Data)
+			}
```
Comment (tsandall) [inline plugins/plugins.go:651]: 👍 It's nice to see this only changed a small amount of code in the manager.

File: server/server.go (hunk 21 above)
Comment (patrick-east) [reply server/server.go:1403]: > When a new bundle downloads and activates do these get invalidated by the server's reload function?

Yep, the prepared rego objects get invalidated, and the manager will update its set of them as well. So the next time we do an eval we will have to create a rego object, which will get the updated ones.

> I assume we're doing this for performance? A short comment explaining why this is done would be useful.

Sort of, as-is the Rego object, when preparing for eval, will look in the store for any resolvers it needs to instantiate unless some were explicitly provided with options. So it does affect performance because it would inadvertently initialize a bunch of wasm vm's when doing the prepare call, which it didn't need to do as we have all of them already initialized on the manager. TL;DR: I'll add a comment 😄

File: internal/wasm/sdk/opa/pool.go (hunk 17 above)
Comment (patrick-east) [reply internal/wasm/sdk/opa/pool.go]: Done

File: internal/wasm/sdk/opa/opa.go (hunk 18 above)
Comment (patrick-east) [reply internal/wasm/sdk/opa/opa.go:215]: Not really, no. IIRC the intention was to allow callers to do more optimized things versus the OPA instance having to like hold onto a map of strings to look them up on each eval. As-is we more easily avoid having to stringify ref's on eval by using the ast.ValueMap to look up the IDs.

Looking into moving it all down a layer into the "OPA" instance.

File: internal/wasm/sdk/opa/vm.go (hunk 19 above)
Comment (patrick-east) [reply internal/wasm/sdk/opa/vm.go:127]: Correct, we have one VM do the parse and then copy the parsed data section of memory into the other vm's.

> though we'll want to get test coverage ASAP
💯 💯

File: internal/wasm/sdk/opa/vm.go (hunk 20 above)
Comment (patrick-east) [reply internal/wasm/sdk/opa/vm.go:554]: Done
